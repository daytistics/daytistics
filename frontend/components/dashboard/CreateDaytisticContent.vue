<template>
    <form @submit.prevent="form.submit">
        <div class="grid gap-4 mb-4 grid-cols-2">
            <div class="col-span-2">
                <label
                    for="type"
                    class="block mb-2 text-sm font-medium text-dark"
                    >What day are we talking about here?</label
                >
                <div class="relative max-w-sm">
                    <div
                        class="absolute inset-y-0 start-0 flex items-center ps-3.5 pointer-events-none w-fit"
                    >
                        <svg
                            class="w-4 h-4 text-day-gray-dark"
                            aria-hidden="true"
                            xmlns="http://www.w3.org/2000/svg"
                            fill="currentColor"
                            viewBox="0 0 20 20"
                        >
                            <path
                                d="M20 4a2 2 0 0 0-2-2h-2V1a1 1 0 0 0-2 0v1h-3V1a1 1 0 0 0-2 0v1H6V1a1 1 0 0 0-2 0v1H2a2 2 0 0 0-2 2v2h20V4ZM0 18a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V8H0v10Zm5-8h10a1 1 0 0 1 0 2H5a1 1 0 0 1 0-2Z"
                            />
                        </svg>
                    </div>
                    <DatePicker
                        @update="form.handleDateUpdate"
                        class="w-fit"
                    />
                </div>
            </div>
        </div>
        <button
            type="submit"
            class="button"
        >
            <svg
                class="me-1 -ms-1 w-5 h-5"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
            >
                <path
                    fill-rule="evenodd"
                    d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
                    clip-rule="evenodd"
                ></path>
            </svg>
            <span>Add</span>
        </button>
    </form>
</template>

<script lang="ts" setup>
import { useToast } from 'vue-toastification';
import useEndpoints from '~/composables/useEndpoints';

const emit = defineEmits(['close']);
const form = useForm();

function useForm() {
    const toast = useToast();
    const date = ref<Date | null>(null);

    const submit = async () => {
        if (!date.value) {
            toast.error('Please select a date');
            return;
        }

        try {
            const daytisticId = (await useEndpoints().createDaytistic(
                date.value as Date
            )) as number;
            await navigateTo(`/dashboard/daytistics/${daytisticId}`);
            toast.success('Daytistic created successfully');
        } catch (error) {
            toast.error('Failed to create daytistic');
        }
    };

    const handleDateUpdate = (newDate: Date) => {
        date.value = newDate;
    };

    return { date, submit, handleDateUpdate };
}
</script>
