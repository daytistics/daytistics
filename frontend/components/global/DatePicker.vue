<template>
    <VueDatePicker
        v-model="date"
        auto-apply
        :max-date="new Date()"
        :enable-time-picker="false"
        :format="date?.toLocaleDateString()"
        @update:model-value="handleUpdate"
    />
</template>

<script lang="ts" setup>
import VueDatePicker from '@vuepic/vue-datepicker';
import '@vuepic/vue-datepicker/dist/main.css';
import '~/assets/css/vue-datepicker.css';

const date = ref<Date | null>(null);

const emit = defineEmits(['update']);

function handleUpdate(newDate: Date) {
    date.value = newDate;
    emit('update', newDate);
}
</script>
