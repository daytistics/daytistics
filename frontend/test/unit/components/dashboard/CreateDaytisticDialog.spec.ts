import { it, describe, expect, vi, beforeEach } from 'vitest';
import { mockNuxtImport, mountSuspended } from '@nuxt/test-utils/runtime';
import { flushPromises } from '@vue/test-utils';
import CreateDaytisticContent from '~~/components/dashboard/CreateDaytisticContent.vue';
import DatePicker from '~~/components/global/DatePicker.vue';

// Create mocks
const mockError = vi.fn();
const mockSuccess = vi.fn();
const mockCreateDaytistic = vi.fn();
const mockNavigateTo = vi.fn();

// Mock vue-toastification
vi.mock('vue-toastification', () => ({
    useToast: () => ({
        error: mockError,
        success: mockSuccess,
    }),
}));

// Mock useEndpoints
vi.mock('~/composables/useEndpoints', () => ({
    default: vi.fn(() => ({
        createDaytistic: mockCreateDaytistic,
    })),
}));

const { navigateTo } = vi.hoisted(() => {
    return {
        navigateTo: vi.fn(),
    };
});

mockNuxtImport('navigateTo', () => navigateTo);

describe('CreateDaytisticContent', () => {
    beforeEach(() => {
        vi.clearAllMocks();
        // Setup successful API response
        mockCreateDaytistic.mockResolvedValue({
            status: 201,
            _data: { id: '123' },
        });
    });

    it('should render the dialog', async () => {
        const wrapper = await mountSuspended(CreateDaytisticContent);
        expect(wrapper.findComponent(DatePicker).exists()).toBe(true);
        expect(wrapper.text()).toContain('What day are we talking about here?');
        expect(wrapper.find('button').text()).toBe('Add');
    });

    it('matches snapshot', async () => {
        const wrapper = await mountSuspended(CreateDaytisticContent);
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows error toast when date is not selected', async () => {
        const wrapper = await mountSuspended(CreateDaytisticContent);

        // Trigger form submission
        await wrapper.find('form').trigger('submit.prevent');
        await flushPromises();

        // Verify toast was called
        expect(mockError).toHaveBeenCalledWith('Please select a date');
        expect(mockCreateDaytistic).not.toHaveBeenCalled();
    });

    it('submits form successfully when date is selected', async () => {
        const wrapper = await mountSuspended(CreateDaytisticContent);
        const testDate = new Date('2024-01-01');

        // First, set the modelValue prop
        await wrapper.findComponent(DatePicker).setValue(testDate);

        // Then emit the update event
        await wrapper.findComponent(DatePicker).vm.$emit('update', testDate);

        // Wait for Vue to process the updates
        await wrapper.vm.$nextTick();

        // Trigger form submission
        await wrapper.find('form').trigger('submit');

        // Wait for all promises to resolve
        await flushPromises();

        // Verify API call was made with correct date
        expect(mockCreateDaytistic).toHaveBeenCalledWith(testDate);

        // FOR AI: Failes with "expected "spy" to be called at least once"
        expect(navigateTo).toHaveBeenCalled();
        expect(mockSuccess).toHaveBeenCalled();
    });

    it('handles API error properly', async () => {
        const wrapper = await mountSuspended(CreateDaytisticContent);
        const testDate = new Date('2024-01-01');
        const errorMessage = 'API Error';

        // Setup error response
        mockCreateDaytistic.mockRejectedValueOnce({
            response: {
                _data: { detail: errorMessage },
            },
        });

        // Simulate date selection and submission
        await wrapper.findComponent(DatePicker).vm.$emit('update', testDate);
        await wrapper.vm.$nextTick();
        await wrapper.find('form').trigger('submit.prevent');
        await flushPromises();

        // Verify error handling
        expect(mockError).toHaveBeenCalled();
        expect(mockNavigateTo).not.toHaveBeenCalled();
    });
});
