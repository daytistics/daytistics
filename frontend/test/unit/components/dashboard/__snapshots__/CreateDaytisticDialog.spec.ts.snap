// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateDaytisticContent > matches snapshot 1`] = `
"<form>
  <div class="grid gap-4 mb-4 grid-cols-2">
    <div class="col-span-2"><label for="type" class="block mb-2 text-sm font-medium text-dark">What day are we talking about here?</label>
      <div class="relative max-w-sm">
        <div class="absolute inset-y-0 start-0 flex items-center ps-3.5 pointer-events-none w-fit"><svg class="w-4 h-4 text-day-gray-dark" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 20 20">
            <path d="M20 4a2 2 0 0 0-2-2h-2V1a1 1 0 0 0-2 0v1h-3V1a1 1 0 0 0-2 0v1H6V1a1 1 0 0 0-2 0v1H2a2 2 0 0 0-2 2v2h20V4ZM0 18a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V8H0v10Zm5-8h10a1 1 0 0 1 0 2H5a1 1 0 0 1 0-2Z"></path>
          </svg></div>
        <div class="dp__main dp__theme_light w-fit" data-datepicker-instance="">
          <div>
            <!---->
            <div class="dp__input_wrap">
              <!----><input data-test="dp-input" class="dp__pointer dp__input_readonly dp__input dp__input_icon_pad dp__input_reg" inputmode="none" placeholder="" autocomplete="off" aria-label="Datepicker input" value="">
              <div>
                <!----><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32" fill="currentColor" aria-hidden="true" class="dp__input_icon dp__input_icons" role="img" aria-label="Calendar icon">
                  <path d="M29.333 8c0-2.208-1.792-4-4-4h-18.667c-2.208 0-4 1.792-4 4v18.667c0 2.208 1.792 4 4 4h18.667c2.208 0 4-1.792 4-4v-18.667zM26.667 8v18.667c0 0.736-0.597 1.333-1.333 1.333 0 0-18.667 0-18.667 0-0.736 0-1.333-0.597-1.333-1.333 0 0 0-18.667 0-18.667 0-0.736 0.597-1.333 1.333-1.333 0 0 18.667 0 18.667 0 0.736 0 1.333 0.597 1.333 1.333z"></path>
                  <path d="M20 2.667v5.333c0 0.736 0.597 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-5.333c0-0.736-0.597-1.333-1.333-1.333s-1.333 0.597-1.333 1.333z"></path>
                  <path d="M9.333 2.667v5.333c0 0.736 0.597 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-5.333c0-0.736-0.597-1.333-1.333-1.333s-1.333 0.597-1.333 1.333z"></path>
                  <path d="M4 14.667h24c0.736 0 1.333-0.597 1.333-1.333s-0.597-1.333-1.333-1.333h-24c-0.736 0-1.333 0.597-1.333 1.333s0.597 1.333 1.333 1.333z"></path>
                </svg>
              </div>
              <!---->
              <!---->
            </div>
          </div>
          <div>
            <transition-stub name="dp-menu-appear-bottom" appear="false" persisted="false" css="true">
              <!---->
            </transition-stub>
          </div>
        </div>
      </div>
    </div>
  </div><button type="submit" class="button"><svg class="me-1 -ms-1 w-5 h-5" fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg">
      <path fill-rule="evenodd" d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z" clip-rule="evenodd"></path>
    </svg><span>Add</span></button>
</form>"
`;

exports[`CreateDaytisticDialog > matches snapshot 1`] = `
"<form>
  <div class="grid gap-4 mb-4 grid-cols-2">
    <div class="col-span-2"><label for="type" class="block mb-2 text-sm font-medium text-dark">What day are we talking about here?</label>
      <div class="relative max-w-sm">
        <div class="absolute inset-y-0 start-0 flex items-center ps-3.5 pointer-events-none w-fit"><svg class="w-4 h-4 text-day-gray-dark" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 20 20">
            <path d="M20 4a2 2 0 0 0-2-2h-2V1a1 1 0 0 0-2 0v1h-3V1a1 1 0 0 0-2 0v1H6V1a1 1 0 0 0-2 0v1H2a2 2 0 0 0-2 2v2h20V4ZM0 18a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V8H0v10Zm5-8h10a1 1 0 0 1 0 2H5a1 1 0 0 1 0-2Z"></path>
          </svg></div>
        <div class="dp__main dp__theme_light w-fit" data-datepicker-instance="">
          <div>
            <!---->
            <div class="dp__input_wrap">
              <!----><input data-test="dp-input" class="dp__pointer dp__input_readonly dp__input dp__input_icon_pad dp__input_reg" inputmode="none" placeholder="" autocomplete="off" aria-label="Datepicker input" value="">
              <div>
                <!----><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32" fill="currentColor" aria-hidden="true" class="dp__input_icon dp__input_icons" role="img" aria-label="Calendar icon">
                  <path d="M29.333 8c0-2.208-1.792-4-4-4h-18.667c-2.208 0-4 1.792-4 4v18.667c0 2.208 1.792 4 4 4h18.667c2.208 0 4-1.792 4-4v-18.667zM26.667 8v18.667c0 0.736-0.597 1.333-1.333 1.333 0 0-18.667 0-18.667 0-0.736 0-1.333-0.597-1.333-1.333 0 0 0-18.667 0-18.667 0-0.736 0.597-1.333 1.333-1.333 0 0 18.667 0 18.667 0 0.736 0 1.333 0.597 1.333 1.333z"></path>
                  <path d="M20 2.667v5.333c0 0.736 0.597 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-5.333c0-0.736-0.597-1.333-1.333-1.333s-1.333 0.597-1.333 1.333z"></path>
                  <path d="M9.333 2.667v5.333c0 0.736 0.597 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-5.333c0-0.736-0.597-1.333-1.333-1.333s-1.333 0.597-1.333 1.333z"></path>
                  <path d="M4 14.667h24c0.736 0 1.333-0.597 1.333-1.333s-0.597-1.333-1.333-1.333h-24c-0.736 0-1.333 0.597-1.333 1.333s0.597 1.333 1.333 1.333z"></path>
                </svg>
              </div>
              <!---->
              <!---->
            </div>
          </div>
          <div>
            <transition-stub name="dp-menu-appear-bottom" appear="false" persisted="false" css="true">
              <!---->
            </transition-stub>
          </div>
        </div>
      </div>
    </div>
  </div><button type="submit" class="button"><svg class="me-1 -ms-1 w-5 h-5" fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg">
      <path fill-rule="evenodd" d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z" clip-rule="evenodd"></path>
    </svg><span>Add</span></button>
</form>"
`;
